runtime/browser/ui/desktop/xwalk_autofill_popup_controller.cc

```diff
@@ -0,0 +1,584 @@
+    return false;
+  }
+}
+
+void XWalkAutofillPopupController::UpdateBoundsAndRedrawPopup() {
+  // TODO(csharp): Since UpdatePopupBounds can change the position of the popup,
+  // the popup could end up jumping from above the element to below it.
+  // It is unclear if it is better to keep the popup where it was, or if it
+  // should try and move to its desired position.
+  UpdatePopupBounds();
+
+  view_->UpdateBoundsAndRedrawPopup();
+}
+
+void XWalkAutofillPopupController::SelectionCleared() {
+  SetSelectedLine(kNoSelection);
+}
+
+void XWalkAutofillPopupController::AcceptSuggestion(size_t index) {
+    const autofill::Suggestion& suggestion = suggestions_[index];
+  delegate_->DidAcceptSuggestion(suggestion.value, suggestion.frontend_id,
+    index);
+}
+
+bool XWalkAutofillPopupController::IsWarning(size_t index) const {
+  return suggestions_[index].frontend_id ==
+      autofill::POPUP_ITEM_ID_WARNING_MESSAGE;
+}
+
+int XWalkAutofillPopupController::GetIconResourceID(
+    const base::string16& resource_name) const {
+  int result = -1;
+  for (size_t i = 0; i < arraysize(kDataResources); ++i) {
+    if (resource_name == base::ASCIIToUTF16(kDataResources[i].name)) {
+      result = kDataResources[i].id;
+      break;
+    }
+  }
+  return result;
+}
+
+void XWalkAutofillPopupController::SetSelectionAtPoint(
+    const gfx::Point& point) {
+  SetSelectedLine(LineFromY(point.y()));
+}
+
+bool XWalkAutofillPopupController::AcceptSelectedLine() {
+  if (selected_line_ == kNoSelection)
+    return false;
+
+  DCHECK_GE(selected_line_, 0);
+  DCHECK_LT(selected_line_, static_cast<int>(GetLineCount()));
+
+  if (!CanAccept(suggestions_[selected_line_].frontend_id))
+    return false;
+
+  AcceptSuggestion(selected_line_);
+  return true;
+}
+
+bool XWalkAutofillPopupController::RemoveSelectedLine() {
+  if (selected_line_ == kNoSelection)
+    return false;
+
+  DCHECK_GE(selected_line_, 0);
+  DCHECK_LT(selected_line_, static_cast<int>(GetLineCount()));
+  return RemoveSuggestion(selected_line_);
+}
+
+gfx::Rect XWalkAutofillPopupController::GetRowBounds(size_t index) {
+  int top = XWalkAutofillPopupView::kPopupBorderThickness;
+  for (size_t i = 0; i < index; ++i) {
+    top += GetRowHeightFromId(suggestions_[i].frontend_id);
+  }
+
+  return gfx::Rect(
+      XWalkAutofillPopupView::kPopupBorderThickness,
+      top,
+      popup_bounds_.width() - 2 *
+        XWalkAutofillPopupView::kPopupBorderThickness,
+      GetRowHeightFromId(suggestions_[index].frontend_id));
+}
+
+void XWalkAutofillPopupController::SetPopupBounds(const gfx::Rect& bounds) {
+  popup_bounds_ = bounds;
+  UpdateBoundsAndRedrawPopup();
+}
+
+bool XWalkAutofillPopupController::IsRTL() const {
+  return popup_controller_->is_rtl();
+}
+
+size_t XWalkAutofillPopupController::GetLineCount() const {
+  return suggestions_.size();
+}
+
+const autofill::Suggestion& XWalkAutofillPopupController::GetSuggestionAt(
+  size_t row) const {
+  return suggestions_[row];
```

Comment:
`[]` would add an element, `at()` should be used or item's presents should have been checked beforehand